### PR TITLE
Add a `VisAugmenter` stats API

### LINKED ISSUE #3836: [Feature Anywhere] Add server-side API to track feature usage
To help track the usage of the new feature, we can add an OSD api call which can utilize the saved object client to fetch things like:
- number of augment-vis saved objects
- number of associations per visualization
- number of visualizations per plugin resource
- aggregate across plugin types
- aggregate across plugin resource types

Because most of the feature usage can be tracked via this saved object type and the values within it, we can collect and aggregate all of that data to organize a useful list of information to show the general usage.

Some other things to consider:
- ownership of the new API(s) - OSD or OSD plugins?
- single vs. multiple APIs

### PR DESCRIPTION
**Pull request by ohltyler**
### Description

Adds a server-side API `/api/vis-augmenter/stats` to track usage of the feature. It does this by collecting all `augment-vis` saved objects, and aggregating values within them. Specifically:
- total number of saved objects
- breakdown and count of unique origin plugins
- breakdown and count of unique plugin resource types
- breakdown and count of plugin resource IDs
- breakdown and count of visualization IDs

This allows us to track what plugins and plugin resource types are being used (e.g., how many anomaly detectors are associated), as well as the individual IDs of the plugin resources and visualizations, to see if there are any unique patterns in how the feature is being used (e.g., many plugin resources to one visualization compared to roughly 1:1 mapping).

To collect these values, this PR also changes the `augment-vis` saved object interface to include all of these fields. 

Additionally, there's some small type optimizations to the saved object loader to process the individual attributes, by introducing `AugmentVisSavedObjectAttributes`. Many of the functions are updated to include this interface.

Existing tests are updated and new tests are added for the functions used in the API call.

### Issues Resolved

closes #3836 

### Check List
- [X] Commits are signed per the DCO using --signoff

### CODE REVIEW

File: src/plugins/vis_augmenter/server/routes/stats.ts
Comment (ashwin-pc) [inline src/plugins/vis_augmenter/server/routes/stats.ts:1]: Can we add documentation for this API? This will be a part of semver so I want us to start documenting the API and its usage so that other plugins can use it easily too.

File: src/plugins/vis_augmenter/common/augment_vis_saved_object_attributes.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  };
+  version: number;
+  // Following fields are optional since they will get set/removed during the extraction/injection
+  // of the vis reference
+  visName?: string;
+  visId?: string;
+  visReference?: {
+    id: string;
+    name: string;
+  };
```
Comment (ashwin-pc) [inline src/plugins/vis_augmenter/common/augment_vis_saved_object_attributes.ts:29]: Can we make a separate interface for these attributes? The convention for `SavedObjectAttributes` is to keep this interface identical to that of the saved object itself. If these properties are not going to be stored in the saved object, lets make a separate interface that extends this instead and use that. Also i dont see this being used anywhere, is it used in the AD plugin? if so can we just create the new interface with these properties there?

File: src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts [hunk 2]
```diff
@@ -24,28 +24,19 @@ export function createSavedAugmentVisClass(services: SavedObjectOpenSearchDashbo
...
-        opts = { id: opts };
-      }
+    constructor(opts: AugmentVisSavedObjectAttributes) {
       super({
         type: SavedAugmentVis.type,
         mapping: SavedAugmentVis.mapping,
         extractReferences,
         injectReferences,
         id: (opts.id as string) || '',
-        indexPattern: opts.indexPattern as IIndexPattern,
```
Comment (ashwin-pc) [inline src/plugins/vis_augmenter/public/saved_augment_vis/_saved_augment_vis.ts:45]: nit: Why was this removed?

File: src/plugins/vis_augmenter/server/routes/stats_helpers.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      (get(visualizationIdMap, visualizationId, 0) as number) + 1;
+  });
+
+  return {
+    total_objs: resp.total,
+    obj_breakdown: {
+      origin_plugin: originPluginMap,
+      plugin_resource_type: pluginResourceTypeMap,
+      plugin_resource_id: pluginResourceIdMap,
+      visualization_id: visualizationIdMap,
```
Comment (ashwin-pc) [inline src/plugins/vis_augmenter/server/routes/stats_helpers.ts:88]: This is just for my understanding, but what do each of these properties mean? like whats the difference between origin_plugin, plugin_resource_type and plugin_resource_id?

File: src/plugins/vis_augmenter/server/routes/stats_helpers.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        ...augmentVisSavedObjectsPage.saved_objects,
+      ];
+    }
+  }
+  return augmentVisSavedObjects;
+};
+
+/**
+ * Given the _find response that contains all of the saved objects, iterate through them and
+ * increment counters for each unique value we are tracking
```
Comment (lezzago) [inline src/plugins/vis_augmenter/server/routes/stats_helpers.ts:57]: We should add some description of what the output exactly means. Maybe providing an example output would help clear things more.

File: src/plugins/vis_augmenter/common/augment_vis_saved_object_attributes.ts (hunk 1 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/common/augment_vis_saved_object_attributes.ts:29]: These are used in the saved object actually - the `visName`, is populated to reference the `visualization` reference in the `references` array, and the `visId` and `visReference` are populated when pulling out the stored vis in `references`. This follows the same convention as used by visualization saved object `extractReferences/injectReferences` logic.

The visualization saved object uses the open-ended `SavedObjectAttributes`, and uses values there that aren't actually indexed in the saved object. I followed the same. From my understanding, the attributes can be values before/after the reference extraction/injection, based on the arguments to those functions themselves.

References:
visualization extract/inject references: [link](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/feature/feature-anywhere/src/plugins/visualizations/public/saved_visualizations/saved_visualization_references.ts)
augment-vis extract/inject references: [link](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/feature/feature-anywhere/src/plugins/vis_augmenter/public/saved_augment_vis/saved_augment_vis_references.ts)

File: src/plugins/vis_augmenter/server/routes/stats_helpers.ts (hunk 3 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/server/routes/stats_helpers.ts:88]: It's mentioned in the description and will be added in the docs, but will add examples here too:
origin_plugin: Anomaly Detection, Alerting
plugin_resource_type: Anomaly Detectors, Alerting Monitors
plugin_resource_id: <detector-1-id>, <monitor-1-id>

File: src/plugins/vis_augmenter/server/routes/stats_helpers.ts (hunk 4 above)
Comment (ohltyler) [reply src/plugins/vis_augmenter/server/routes/stats_helpers.ts:57]: So this is defined within the `resp` argument type - `SavedObjectsFindResponse`. Prefer to not add an example here which could become outdated. By looking up the defined type, it will show the fields.

Also, the saved object list is located under `saved_objects` field as seen on line 67

**Review by ashwin-pc — APPROVED** [review]

**Review by lezzago — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]
